### PR TITLE
Issue #4124 - autobahn tests for jetty, javax websocket APIs

### PR DESCRIPTION
**Pull request by lachlan-roberts**
**Issue #4124**
Added autobahn tests to allow testing of the jetty and javax APIs instead of only core.
Also fixed some problems stopping the tests from passing.

- No longer selecting duplicate extensions in the javax `ContainerDefaultConfigurator`.
- Correctly copy payloads for ping frames in jetty & javax FrameHandlers.
- Add `ByteBuffer` to javadoc for `@OnWebSocketMessage` annotation in jetty API.
- Cleaned up some test logging for `EventSocket`.
- Add `autoFragment` and `maxFrameSize` settings to `WebSocketPolicy`.
- Fix early validation for multiple extensions in `ServletUpgradeResponse.setExtensions`.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

### CODE REVIEW

**Comment by lachlan-roberts** [comment]
@gregw Not in its current state.

The javax autobahn tests are failing because of #4148 / #4152. Once this is fixed it would be possible to run the autobahn tests for all 3 APIs on jenkins with a few changes, but only if we exclude the `permessage-deflate` server tests (excluding them you can run all the tests in a few seconds otherwise it can take over 20min).

**Comment by olamy** [comment]
@lachlan-roberts those autobahn tests need to be executed via command line. That's not related to the current maven plugin we use?

**Comment by lachlan-roberts** [comment]
@olamy Yeah I was just running these ones on the command line for now, I think the current maven plugin is still only using the core API. Do you think it would be possible to eventually set up all three APIs to run test on jenkins for both the server and client tests?

File: jetty-websocket/javax-websocket-common/src/main/java/org/eclipse/jetty/websocket/javax/common/JavaxWebSocketFrameHandler.java [hunk 1]
```diff
@@ -545,13 +545,7 @@ public void onClose(Frame frame, Callback callback)
...
     public void onPing(Frame frame, Callback callback)
     {
-        ByteBuffer payload = BufferUtil.EMPTY_BUFFER;
-
-        if (frame.hasPayload())
-        {
-            payload = ByteBuffer.allocate(frame.getPayloadLength());
-            BufferUtil.put(frame.getPayload(), payload);
-        }
+        ByteBuffer payload = BufferUtil.copy(frame.getPayload());
```
Comment (joakime) [inline jetty-websocket/javax-websocket-common/src/main/java/org/eclipse/jetty/websocket/javax/common/JavaxWebSocketFrameHandler.java]: This is different then the original version.
The original version maintained the ByteBuffer, even if null/empty (it would use BufferUtil.EMPTY_BUFFER)

File: jetty-websocket/javax-websocket-server/src/main/java/org/eclipse/jetty/websocket/javax/server/config/ContainerDefaultConfigurator.java [hunk 2]
```diff
@@ -78,7 +80,16 @@ public boolean checkOrigin(String originHeaderValue)
...
+        List<Extension> negotiatedExtensions = new ArrayList<>();
+        for (Extension ext : requested)
+        {
+            // Only choose the first extension if multiple with the same name.
+            long matches = negotiatedExtensions.stream().filter(e -> e.getName().equals(ext.getName())).count();
+            if (matches == 0)
+                negotiatedExtensions.add(ext);
+        }
+
+        return negotiatedExtensions;
```
Comment (joakime) [inline jetty-websocket/javax-websocket-server/src/main/java/org/eclipse/jetty/websocket/javax/server/config/ContainerDefaultConfigurator.java]: This isn't right.

A few points ...

* Multiple copies of the same extension are allowed. Eg: `identity; step=1, permessage-deflate, identity; step=2`
* This API should validate each `requested` name against the list in `installed` and not return `requested` that does not show up in `installed`.

File: jetty-websocket/jetty-websocket-common/src/test/java/org/eclipse/jetty/websocket/common/OutgoingMessageCapture.java [hunk 3]
```diff
@@ -295,6 +295,18 @@ public long getMaxTextMessageSize()
                 return maxMessageSize;
             }
 
+            @Override
+            public long getMaxFrameSize()
+            {
+                return 0;
```
Comment (joakime) [inline jetty-websocket/jetty-websocket-common/src/test/java/org/eclipse/jetty/websocket/common/OutgoingMessageCapture.java]: What does a maxFrameSize of 0 mean?

File: jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java [hunk 4]
```diff
@@ -50,44 +49,25 @@ public WebSocketBehavior getBehavior()
         return WebSocketBehavior.SERVER;
     }
 
-    /**
-     * If true, frames are automatically fragmented to respect the maximum frame size.
-     *
-     * @return whether to automatically fragment incoming WebSocket Frames.
-     */
+    @Override
```
Comment (joakime) [inline jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java]: Why did you remove the javadoc?
That just means the javadoc will not be present on this public class now.

File: jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java [hunk 5]
```diff
@@ -50,44 +49,25 @@ public WebSocketBehavior getBehavior()
...
     {
         return factory.isAutoFragment();
     }
 
-    /**
-     * If set to true, frames are automatically fragmented to respect the maximum frame size.
-     *
-     * @param autoFragment whether to automatically fragment incoming WebSocket Frames.
-     */
+    @Override
```
Comment (joakime) [inline jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java]: Restore javadoc

File: jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java [hunk 6]
```diff
@@ -50,44 +49,25 @@ public WebSocketBehavior getBehavior()
...
     {
         factory.setAutoFragment(autoFragment);
     }
 
-    /**
-     * The maximum payload size of any WebSocket Frame which can be received.
-     *
-     * @return the maximum size of a WebSocket Frame.
-     */
+    @Override
```
Comment (joakime) [inline jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java]: Restore javadoc

File: jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java [hunk 7]
```diff
@@ -50,44 +49,25 @@ public WebSocketBehavior getBehavior()
...
 
-    /**
-     * The maximum payload size of any WebSocket Frame which can be received.
-     * <p>
-     * WebSocket Frames over this maximum will result in a close code 1009 {@link StatusCode#MESSAGE_TOO_LARGE}
-     * </p>
-     *
-     * @param maxFrameSize the maximum allowed size of a WebSocket Frame.
-     */
+    @Override
```
Comment (joakime) [inline jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java]: Restore javadoc

File: jetty-websocket/javax-websocket-server/src/main/java/org/eclipse/jetty/websocket/javax/server/config/ContainerDefaultConfigurator.java (hunk 2 above)
Comment (joakime) [reply jetty-websocket/javax-websocket-server/src/main/java/org/eclipse/jetty/websocket/javax/server/config/ContainerDefaultConfigurator.java]: The `permessage-deflate` extension has special language that it reserves the RSV1 bit.
So if you have a offered extensions of `identity; step=1, permessage-deflate, identity; step=2, permessage-deflate; client_no_context_takeover`, then what happens is ...

1. `identity; step=1` is initialized successfully.
2. `permessage-deflate` is initialized, and reserves the RSV1 bit. successfully.
3. `identity; step=2` is initialized successfully.
4. `permessage-deflate; client_no_context_takeover` is initialized, and _fails_ to reserve the RSV1 bit, throwing an exception, and erroring out the handshake. (the language used in RFC6455 is `MUST immediately _Fail the WebSocket Connection_`)

File: jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java (hunk 4 above)
Comment (lachlan-roberts) [reply jetty-websocket/jetty-websocket-server/src/main/java/org/eclipse/jetty/websocket/server/JettyWebSocketServletFactory.java]: I have not removed the javadoc just moved it up to `WebSocketPolicy`, the javadoc will still be there because of the `@Override`

File: jetty-websocket/javax-websocket-common/src/main/java/org/eclipse/jetty/websocket/javax/common/JavaxWebSocketFrameHandler.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-websocket/javax-websocket-common/src/main/java/org/eclipse/jetty/websocket/javax/common/JavaxWebSocketFrameHandler.java]: It actually does the same thing as the original version because `frame.getPayload()` will return `BufferUtil.EMPTY_BUFFER` if it has a null payload.

File: jetty-websocket/jetty-websocket-common/src/test/java/org/eclipse/jetty/websocket/common/OutgoingMessageCapture.java (hunk 3 above)
Comment (lachlan-roberts) [reply jetty-websocket/jetty-websocket-common/src/test/java/org/eclipse/jetty/websocket/common/OutgoingMessageCapture.java]: It just means we don't enforce the maxFrameSize. But here this is in the method `getFakeSession()` so this is probably not even used here.

File: jetty-websocket/javax-websocket-server/src/main/java/org/eclipse/jetty/websocket/javax/server/config/ContainerDefaultConfigurator.java (hunk 2 above)
Comment (lachlan-roberts) [reply jetty-websocket/javax-websocket-server/src/main/java/org/eclipse/jetty/websocket/javax/server/config/ContainerDefaultConfigurator.java]: I don't think this is how it works, I'm not so sure with the `identity` extension case, but I know that with `permessage-deflate` the client can offer the same extension multiple times with different parameters and the server can choose the one to use. It would not result in a failed handshake.

[RFC6455 Section 9.1](https://tools.ietf.org/html/rfc6455#section-9.1)
```
Any interactions
between multiple extensions MAY be defined in the documents defining
the extensions.  In the absence of such definitions, the
interpretation is that the header fields listed by the client in its
request represent a preference of the header fields it wishes to use,
with the first options listed being most preferable.
```

[RFC7692 'Compression Extensions' Section 5](https://tools.ietf.org/html/rfc7692#section-5)
```
A client may also offer multiple
PMCE choices to the server by including multiple elements in the
"Sec-WebSocket-Extensions" header, one for each PMCE offered.  This
set of elements MAY include multiple PMCEs with the same extension
name to offer the possibility to use the same algorithm with
different configuration parameters.
```
